### PR TITLE
Feat: Shadow block recovery mechanism

### PR DESCRIPTION
**Pull request by jcnelson**
This introduces the notion of "shadow blocks" and "shadow tenures" as a means of recovering from a chain stall.  In the event of an irrecoverable chain stall, we'd pass an emergency SIP to declare that one or more empty sortitions actually commit to a "shadow block" that contains whatever state is needed to get the chain un-stuck.  All nodes would ship with copies of these shadow blocks, and would process them the same as any other kind of block (except that they could not be mined, uploaded, or downloaded).

Leaving as a draft for now until I can expand test coverage.  This touches a lot of consensus-critical code paths.

----
Here is a summary of consensus changes for shadow blocks:
* All shadow blocks _MUST_ have a `version` whose highest bit is set.
* If any block in a tenure is a shadow block, then all blocks in that tenure _MUST_ be shadow blocks (it's a "shadow tenure")
* No non-shadow blocks can be inserted into a shadow tenure.
* The block-commit of the first child tenure of a shadow block tenure _MUST_ set `parent_vtxindex` to 0, which is the burnchain coinbase transaction.  We assume that shadow blocks all descend from their respective PoX anchor blocks, so this will _not_ be checked for the first child tenure.  Instead, the block's parent sortition will be checked to verify that it is a shadow tenure and that the block-commit `parent_vtxindex` is 0 when validating the child block.
* The VRF proof of a shadow block will not be validated (since there's no seed to use)
* The miner signature of a shadow block will not be validated (since there's no miner)
* For the purposes of finding the PoX anchor block, a shadow tenure in the prepare-phase always counts as a sortition, even if there were no block-commits in that burnchain block.
* The signing weight of a shadow block is the maximal weight of the reward set.  It's as if all signers signed it.
* The coinbase of a shadow block _MUST_ be the burn address for the network.  Shadow blocks will produce STX, but no one will receive it.
* The memo field of the coinbase of a shadow block _MUST_ be the hash of the tenure-start block of the shadow block's parent tenure.  This is required because shadow blocks do not have corresponding block-commits.  Instead, we capture the tenure start block hash that would have gone into the block-commit as the memo field in the shadow block coinbase.
* Shadow blocks do not increase the total BTC spend for the chain
* The VRF proof for the child of a shadow block must come from the VRF seed of the shadow block -- i.e. the hash of the VRF proof in the parent shadow block's tenure's coinbase.

### CODE REVIEW

**Review by obycode — COMMENTED** [review]
Looking good!

**Comment by jcnelson** [comment]
N.B. This PR needs to be followed by (or extend with) a PR to re-mine a sequence of blocks as shadow blocks, so we can fix a consensus bug on testnet.

We'll need more elaborate chain replay logic in the future to preserve causal relationships between affected transactions if possible.  At a minimum today, we can preserve the VRF proofs.

**Comment by jcnelson** [comment]
Okay, this PR is getting kinda big.  I'm going to hold off on adding the tooling for generating shadow blocks for the time being.  I'll add that as a separate PR so we can move this one along.

File: testnet/stacks-node/src/config.rs
Comment (kantai) [inline testnet/stacks-node/src/config.rs:1971]: What is the impact of this change? Does it impact mainnet? Is this config option only ever evaluated in testnet? Did testnet node's previously default to ignoring affirmation maps? Does this only impact tests?

Comment (jcnelson) [reply testnet/stacks-node/src/config.rs:1965]: > What is the impact of this change? 

When booting a node in epoch 2.05 or earlier, we rely on the affirmation map to tell us whether or not anchor blocks exist as determined by block-commits.  We know for a fact that in all testnets and mainnet, every reward cycle either has an anchor block as determined by the block-commit history or no PoX anchor block was chosen.

This had not been the case prior to resetting primary testnet to use Bitcoin regtest.

> Is this config option only ever evaluated in testnet?

Nope, it applies on mainnet too.

> Did testnet node's previously default to ignoring affirmation maps? 

Yes, for epochs prior to 2.1.

> Does this only impact tests?

I had to set this to `true` to get my mainline testnet node booted up.

File: stackslib/src/chainstate/nakamoto/miner.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/miner.rs:1]: Why does the BlockBuilder need to support shadow blocks? Do shadow blocks get built by the block builder?

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/mod.rs:1]: I'm kind of wary of how shadow blocks are handled in this module.

This code adds some special cases to various normal block execution paths to allow "special" behaviors in the event the block being evaluated is a shadow block. These end up peppering `check_coinbase_tx`, `validate_against_burnchain`, `setup_block`, and other functions with a bunch of `if-else` statements switching on the block's shadow block status, which ends up propagating to several variables by changing non-option types to option types and moving pretty large blocks of validation into `else` branches. While I don't see there being an obvious logic problem here, this structure does make the normal block evaluation's control flow more complex and increases the space for errors.

If possible, I think it might be better to have something like a separate path for the shadow blocks which is switched into early on in the control flow (e.g., separate `setup_block` functions). If there's opportunities for DRY'ing the code between those functions, great, but the hope is that it wouldn't end up duplicating much.

File: stackslib/src/burnchains/tests/mod.rs
Comment (jferrant) [inline stackslib/src/burnchains/tests/mod.rs:460]: Nit: would be good probably to add a message here for the failure case like "Parent block is a shadow block. There should be no block commit." or something like that.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/mod.rs:1555]: You could simplify these two if statements into one I do believe by just doing

```rust
                // must be the standard burn address for this network
                let burn_addr = StacksAddress::burn_address(mainnet).to_account_principal();
                if Some(burn_addr) != recipient_opt {
                    warn!("Invalid shadow block: recipient does not burn");
                    return Err(ChainstateError::InvalidStacksBlock(
                        "Shadow block did not pay to burn address".into(),
                    ));
                }
```

File: stackslib/src/chainstate/nakamoto/staging_blocks.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/staging_blocks.rs:538]: Is it possible to have some unit tests for the database calls?

Not sure how easy that is to set up, but would be nice.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs:1]: Sure, I'll refactor this and put all shadow-block logic into its own module and minimize the number of new `if-else` branches created (and if they do need to be added, I'll keep their bodies short -- such as limiting them to a method call).

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs]: This has been simplified by refactoring the shadow block system in bdd5ea3

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs:1]: This has been addressed in bdd5ea3

File: stackslib/src/burnchains/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/burnchains/tests/mod.rs]: Done

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/mod.rs:4118]: I think this may be a github artifact from merging #5291

**Comment by jcnelson** [comment]
@kantai @jferrant @obycode This is ready for re-review.  I'm going to work on tooling for (quickly) synthesizing shadow blocks via `stacks-inspect` so we can test this on the Nakamoto testnet.

File: stackslib/src/net/download/nakamoto/tenure_downloader_unconfirmed.rs
Comment (obycode) [inline stackslib/src/net/download/nakamoto/tenure_downloader_unconfirmed.rs:809]: Do we need an issue for this TODO?

Comment (jcnelson) [reply stackslib/src/net/download/nakamoto/tenure_downloader_unconfirmed.rs:809]: It already exists: https://github.com/stacks-network/stacks-core/issues/5268

**Comment by jcnelson** [comment]
Look at that.

All of that.

All those green :heavy_check_mark:'s in CI.

Outstanding work on making them all reliable @jferrant!  It drastically improves our confidence that we can make consensus-critical changes like this.  Well done!

**Review by kantai — DISMISSED** [review]

File: stackslib/src/chainstate/nakamoto/shadow.rs
Comment (obycode) [inline stackslib/src/chainstate/nakamoto/shadow.rs:488]: Could this and `make_shadow_tenure` be gated to ensure it isn't called by the node? I assume you'll need this for the tooling (haven't started reviewing that PR yet), otherwise it could be gated just for tests.

File: stackslib/src/chainstate/nakamoto/shadow.rs
Comment (obycode) [inline stackslib/src/chainstate/nakamoto/shadow.rs:570]:
```suggestion
    /// The details and circumstances will be recorded in an accompanying SIP.
```

File: stackslib/src/net/tests/download/nakamoto.rs
Comment (obycode) [inline stackslib/src/net/tests/download/nakamoto.rs:2529]:
```suggestion
/// Test booting up a node where there is one shadow block in the prepare phase, as well as some
```

**Review by obycode — COMMENTED** [review]
This looks great! Just had some minor comments.

File: stackslib/src/chainstate/nakamoto/miner.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/miner.rs:404]: I think it would be cleaner if you did something like this:
```rust
        let Some(tenure_block_commit) = info.tenure_block_commit_opt.is_none() else {
            return Err(Error::InvalidStacksBlock(
                "Block-commit is required; cannot mine a shadow block".into(),
            ));
        }
 ```
 and then got rid of the extra unwrap below.
 Unless of some borrowing specific reason you did it this way.

**Review by jferrant — APPROVED** [review]

File: stackslib/src/chainstate/nakamoto/shadow.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/shadow.rs:488]: It's needed by the tooling.

File: stackslib/src/net/download/nakamoto/tenure_downloader.rs
Comment (jferrant) [inline stackslib/src/net/download/nakamoto/tenure_downloader.rs:717]: Does this need a ticket?

**Review by jferrant — DISMISSED** [review]

File: stackslib/src/net/api/setloglevel.rs
Comment (jferrant) [inline stackslib/src/net/api/setloglevel.rs:132]: Does this TODO need an issue?

File: stackslib/src/net/api/setloglevel.rs
Comment (jferrant) [inline stackslib/src/net/api/setloglevel.rs:1]: This seems very unrelated to this PR?

Comment (jferrant) [reply stackslib/src/net/api/setloglevel.rs:1]: Also I think we should update the ChangeLog to include it if we do plan to merge this as part of this PR.

**Review by obycode — APPROVED** [review]

**Review by obycode — DISMISSED** [review]

**Comment by jcnelson** [comment]
ping @jferrant @hstove @kantai

File: stackslib/src/chainstate/nakamoto/shadow.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/shadow.rs:122]: These comments seem contradictory to what the function does unless I am missing something. It says we check the vrf proof was generated by this block's miner, but then we say we can't verify the VRF proof because the creator of the shadow block isn't a miner? I think its really that we just ensure that there IS a vrf_proof but do not actually do anything else with it. Side Nit: there is a typo "parnet" -> "parent"

File: stackslib/src/net/api/getsortition.rs
Comment (hstove) [inline stackslib/src/net/api/getsortition.rs:94]: Can we also add a `is_shadow` to `SortitionInfo`?

**Review by hstove — CHANGES_REQUESTED** [review]
Overall LGTM! I have some questions and a few places where I think a small change might be appropriate.

**Comment by jcnelson** [comment]
ping @hstove @jferrant @obycode

**Review by obycode — APPROVED** [review]

File: stackslib/src/burnchains/mod.rs
Comment (hstove) [reply stackslib/src/burnchains/mod.rs:474]: Are you referring to this: https://api.nakamoto.testnet.hiro.so/v2/pox

If so, I think the unlock blocks are still off?

File: stackslib/src/chainstate/burn/operations/leader_block_commit.rs
Comment (hstove) [reply stackslib/src/chainstate/burn/operations/leader_block_commit.rs:1050]: Oh woops, I meant should this line also have `&& self.parent_block_ptr != 0`

File: stackslib/src/burnchains/mod.rs
Comment (hstove) [reply stackslib/src/burnchains/mod.rs:474]: Also different than https://api.nakamoto-4.stg.hiro.so/v2/pox if that's what you are referring to

Comment (hstove) [reply stackslib/src/burnchains/mod.rs:474]: Also, I see now that this is only really used in the CLI tools for shadow blocks, so maybe those specific values don't really matter. If that's the case, we can just resolve this.

**Comment by hstove** [comment]
@jcnelson I responded to the [one unresolved comment](https://github.com/stacks-network/stacks-core/pull/5296#discussion_r1854227436) - once that's fixed I think this is good to merge!

File: stackslib/src/chainstate/burn/operations/leader_block_commit.rs
Comment (jcnelson) [reply stackslib/src/chainstate/burn/operations/leader_block_commit.rs:1050]: I don't think so, because this executes inside of an if-statement whose predicate is `self.parent_block_ptr != 0 || self.parent_vtxindex != 0`.  If `maybe_shadow_block` is `true`, then `self.parent_vtxindex == 0`, so `self.parent_block_ptr != 0`.

**Comment by jcnelson** [comment]
Thanks @hstove!  Replied.

**Review by hstove — APPROVED** [review]

**Pull request merged**
